src/nspawn/nspawn.c

```diff
@@ -4415,16 +4416,30 @@ static int run_container(
         }
 
         if (arg_private_network) {
+                _cleanup_close_ int current_netns_fd = -1;
+
                 if (!arg_network_namespace_path) {
                         /* Wait until the child has unshared its network namespace. */
                         if (!barrier_place_and_sync(&barrier)) /* #3 */
                                 return log_error_errno(SYNTHETIC_ERRNO(ESRCH), "Child died too early");
                 }
 
-                r = move_network_interfaces(*pid, arg_network_interfaces);
+                /* Even if the child exits, the child's network namespace namespace will stay alive
+                 * until this file descriptor is closed. */
+                *child_netns_fd = open_network_namespace(*pid);
+                if (*child_netns_fd < 0)
+                        return *child_netns_fd;
+
+                current_netns_fd = open_network_namespace(getpid());
+                if (current_netns_fd < 0)
+                        return current_netns_fd;
+
+                r = move_network_interfaces(current_netns_fd, *child_netns_fd, arg_network_interfaces);
                 if (r < 0)
                         return r;
 
+                *network_interfaces_moved = true;
+
```

Comment:
we generally try to avoid functions that clobber their return values even on failure, see coding style.

That said, I think you should move the ifaces back inside of run_container() instead of in run(), to properly deal with reboots of the container. i.e. we call run_container() in a loop as long as the contains is restarted via reboot(), and you probably want to move back the ifaces from the guest ns to the host on each cycle of that, to move them again into the new netns of the next reboot instance. But that also means you don't need to return the netns here, and hence the argument clobbering goes away too

Reply:
Moved everything into `run_container`.